### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by jgersti**
This is a edit (and corrected) version of some general purpose refactoring and char saving measures.
This accomplished by heavy usage of the operators `?.` and `??` and removal of obsolete or unnecessary code.

edit:
master: 87374 chars
Refactor: 85584 chars

### CODE REVIEW

File: EasyCommands/BlockHandlers/TextSurfaceHandlers.cs [hunk 1]
```diff
@@ -63,11 +63,11 @@ public class TextSurfaceHandler : MultiInstanceBlockHandler<IMyTextSurface> {
             public override string Name(IMyTextSurface block) => block.DisplayName;
 
             public override IEnumerable<IMyTextSurface> GetInstances(IMyTerminalBlock block) =>
-                block is IMyTextSurface ?
-                Enumerable.Repeat((IMyTextSurface)block, 1) :
-                block is IMyTextSurfaceProvider ?
-                Enumerable.Range(0, ((IMyTextSurfaceProvider)block).SurfaceCount).Select(i => ((IMyTextSurfaceProvider)block).GetSurface(i)) :
-                Enumerable.Empty<IMyTextSurface>();
+                block is IMyTextSurface
+                    ? Once((IMyTextSurface)block)
```
Comment (MerlinofMines) [inline EasyCommands/BlockHandlers/TextSurfaceHandlers.cs:67]: nit: I think you could use NewList() here to save a few chars

File: EasyCommands/BlockHandlers/TextSurfaceHandlers.cs [hunk 2]
```diff
@@ -63,11 +63,11 @@ public class TextSurfaceHandler : MultiInstanceBlockHandler<IMyTextSurface> {
...
+                    : block is IMyTextSurfaceProvider
+                        ? Range(0, ((IMyTextSurfaceProvider)block).SurfaceCount).Select(((IMyTextSurfaceProvider)block).GetSurface)
+                        : Empty<IMyTextSurface>();
```
Comment (MerlinofMines) [inline EasyCommands/BlockHandlers/TextSurfaceHandlers.cs:70]: nit: NewList<IMyTextSurface>()

File: EasyCommands/CommandParsers/ParameterParsers.cs [hunk 3]
```diff
@@ -444,30 +444,28 @@ partial class Program {
...
-        Token[] TokenizeEnclosed(String token, string characters, Func<String, Token[]> parseSubTokens) =>
+        IEnumerable<Token> TokenizeEnclosed(String token, string characters, Func<String, IEnumerable<Token>> parseSubTokens) =>
             characters.Length == 0 ? parseSubTokens(token) :
                 token.Trim().Split(characters[0])
                 .SelectMany((element, index) => index % 2 == 0  // If even index
-                    ? TokenizeEnclosed(element, characters.Remove(0,1), parseSubTokens)  // Split the item
-                    : new Token[] { new Token(element, true, characters.Length > 1) })  // Keep the entire item
-                .ToArray();
+                    ? TokenizeEnclosed(element, characters.Remove(0, 1), parseSubTokens)  // Split the item
+                    : Once(new Token(element, true, characters.Length > 1)));  // Keep the entire item
```
Comment (MerlinofMines) [inline EasyCommands/CommandParsers/ParameterParsers.cs:456]: I see Once() several times. Would NewList() work here to save 3 chars per usage?

File: EasyCommands/BlockHandlers/TextSurfaceHandlers.cs (hunk 1 above)
Comment (MerlinofMines) [reply EasyCommands/BlockHandlers/TextSurfaceHandlers.cs:67]: Rather, why is the Once method necessary, given ```NewList(item)``` ?

File: EasyCommands/Common/Utilities.cs [hunk 4]
```diff
@@ -30,10 +30,16 @@ partial class Program {
...
+        public static IEnumerable<int> Range(int start, int end) => Enumerable.Range(start, end);
+        public static IEnumerable<T> Empty<T>() => Enumerable.Empty<T>();
+        public static IEnumerable<T> Once<T>(T element) => Enumerable.Repeat(element, 1);
 
         //Other useful utilities
-        public static Variable GetStaticVariable(object o) => new StaticVariable(ResolvePrimitive(o));
+        public static IVariable GetStaticVariable(object o) => new StaticVariable(ResolvePrimitive(o));
+        public static IVariable EmptyList() => GetStaticVariable(NewKeyedList());
         public static Vector3D Vector(double x, double y, double z) => new Vector3D(x, y, z);
+        public static bool AnyNotNull(params Object[] objects) => objects.Any(o => o != null);
```
Comment (MerlinofMines) [inline EasyCommands/Common/Utilities.cs:43]: We only use this method a couple of times, is it less chars than checking the nulls directly?

File: EasyCommands/Commands/Commands.cs [hunk 5]
```diff
@@ -260,27 +260,27 @@ public class SendCommand : Command {
...
                 return true;
             }
         }
 
         public class TransferItemCommand : Command {
-            public Selector from;//Must be Inventory
-            public Selector to;//Must be Inventory
-            public Variable first, second;//One of these is an amount (nullable), other must be ItemFilter (non nullable)
+            public ISelector from;//Must be Inventory
+            public ISelector to;//Must be Inventory
```
Comment (MerlinofMines) [inline EasyCommands/Commands/Commands.cs:280]: nit: ```public ISelector from, to;```

File: EasyCommands/Commands/Commands.cs [hunk 6]
```diff
@@ -412,7 +410,7 @@ public class MultiActionCommand : Command {
                     }
                 }
 
-                if (currentCommands != null && currentCommands.Count > 0) return false;
+                if (currentCommands?.Count > 0) return false;
```
Comment (MerlinofMines) [inline EasyCommands/Commands/Commands.cs:413]: is the ``?? 0``` needed here as well? Otherwise, is it not needed above?

File: EasyCommands/Common/Aggregators.cs [hunk 7]
```diff
@@ -20,6 +20,6 @@
 namespace IngameScript {
     partial class Program {
         public delegate Primitive Aggregator(IEnumerable<object> blocks, Func<object,Primitive> primitiveSupplier);
-        public Aggregator SumAggregator = (blocks, primitiveSupplier) => blocks.Select(b => primitiveSupplier(b)).DefaultIfEmpty(ResolvePrimitive(0)).Aggregate((a, b) => a.Plus(b));
+        public Aggregator SumAggregator = (blocks, primitiveSupplier) => blocks.Select(primitiveSupplier).Aggregate((Primitive)null, (a, b) => a?.Plus(b) ?? b) ?? ResolvePrimitive(0);
```
Comment (MerlinofMines) [inline EasyCommands/Common/Aggregators.cs:23]: Why was this change needed?

File: EasyCommands/Common/Properties.cs [hunk 8]
```diff
@@ -45,19 +45,19 @@ public class PropertySupplier {
...
-            PropertySupplier ResolvePropertyType(BlockHandler blockHandler, Return? defaultType = null) {
+            PropertySupplier ResolvePropertyType(IBlockHandler blockHandler, Return? defaultType = null) {
                 if (propertyType != null) return this;
-                if (direction.HasValue) return blockHandler.GetDefaultProperty(direction.Value);
-                if (propertyValue != null) return blockHandler.GetDefaultProperty(propertyValue.GetValue().returnType);
-                if (defaultType.HasValue) return blockHandler.GetDefaultProperty(defaultType.Value);
+                if (direction != null) return blockHandler.GetDefaultProperty(direction.Value);
+                var returnType = propertyValue?.GetValue().returnType ?? defaultType;
+                if (returnType != null) return blockHandler.GetDefaultProperty(returnType.Value);
                 return blockHandler.GetDefaultProperty(blockHandler.GetDefaultDirection());
```
Comment (MerlinofMines) [inline EasyCommands/Common/Properties.cs:61]: nit:  

```
return returnType != null
? blockHandler.GetDefaultProperty(returnType.Value);
: blockHandler.GetDefaultProperty(blockHandler.GetDefaultDirection());
```

File: EasyCommands/Common/Utilities.cs [hunk 9]
```diff
@@ -30,10 +30,16 @@ partial class Program {
         public static List<T> NewList<T>(params T[] elements) => new List<T>(elements);
         public static Dictionary<T, U> NewDictionary<T, U>(params KeyValuePair<T, U>[] elements) => elements.ToDictionary(e => e.Key, e => e.Value);
         public static KeyValuePair<T, U> KeyValuePair<T, U>(T key, U value) => new KeyValuePair<T, U>(key, value);
-        public static Variable EmptyList() => GetStaticVariable(NewKeyedList());
+
+        //Utilities for constructing enumerables
+        public static IEnumerable<int> Range(int start, int end) => Enumerable.Range(start, end);
+        public static IEnumerable<T> Empty<T>() => Enumerable.Empty<T>();
+        public static IEnumerable<T> Once<T>(T element) => Enumerable.Repeat(element, 1);
```
Comment (MerlinofMines) [inline EasyCommands/Common/Utilities.cs:37]: not sure I understand the value add from these ones.  How much better is the performance?  These two cost around 100 chars.

nit: can we remove ```Enumerable.```, as we did elsewhere?

File: EasyCommands/BlockHandlers/TextSurfaceHandlers.cs (hunk 1 above)
Comment (jgersti) [reply EasyCommands/BlockHandlers/TextSurfaceHandlers.cs:67]: no allocation that gets thrown away immediatly since that expression is expanded in a `SelectMany` anyway.

File: EasyCommands/BlockHandlers/TextSurfaceHandlers.cs (hunk 2 above)
Comment (jgersti) [reply EasyCommands/BlockHandlers/TextSurfaceHandlers.cs:70]: see above

File: EasyCommands/CommandParsers/ParameterParsers.cs (hunk 3 above)
Comment (jgersti) [reply EasyCommands/CommandParsers/ParameterParsers.cs:456]: see above

File: EasyCommands/Common/Utilities.cs (hunk 4 above)
Comment (jgersti) [reply EasyCommands/Common/Utilities.cs:43]: if i havent miscounted it is the same right now.

File: EasyCommands/Commands/Commands.cs (hunk 5 above)
Comment (jgersti) [reply EasyCommands/Commands/Commands.cs:280]: :)

File: EasyCommands/Commands/Commands.cs (hunk 6 above)
Comment (jgersti) [reply EasyCommands/Commands/Commands.cs:413]: `null > 0` is `false` so giving a sensible default is not necessary here.

File: EasyCommands/Common/Aggregators.cs (hunk 7 above)
Comment (jgersti) [reply EasyCommands/Common/Aggregators.cs:23]: it saves chars. it may also enable us to deliver a appropriate default value vor non numeric inputs in the future.

File: EasyCommands/Common/Utilities.cs (hunk 9 above)
Comment (jgersti) [reply EasyCommands/Common/Utilities.cs:37]: The main purpose of these functions is to better signal intent and avoidance of allocations. Some but not all usage can be replaced with `NewList`. All in all adding these functions costs only around 20-30 chars but increases readability and removes allocations.

Instead of `Enumerable.` these functions are called, so no

File: EasyCommands/BlockHandlers/TextSurfaceHandlers.cs (hunk 1 above)
Comment (jgersti) [reply EasyCommands/BlockHandlers/TextSurfaceHandlers.cs:67]: also the branches inside `?:` must have the same type

File: EasyCommands/CommandParsers/ParameterParsers.cs [hunk 10]
```diff
@@ -402,31 +402,27 @@ partial class Program {
...
-                commandParameters.Add(new AmbiguousStringCommandParameter(token.original, false, subtokenParams.ToArray()));
-            } else if (propertyWords.ContainsKey(t)) {
-                commandParameters.AddList(propertyWords[t]);
-            } else { //If no property matches, must be a string
+            else if (token.isString)
+                commandParameters.Add(new AmbiguousStringCommandParameter(token.original, false, ParseCommandParameters(Tokenize(token.token)).ToArray()));
+            else if (propertyWords.ContainsKey(token.token))
+                commandParameters.AddList(propertyWords[token.token]);
+            else //If no property matches, must be a string
                 commandParameters.Add(new AmbiguousStringCommandParameter(token.original, true));
```
Comment (MerlinofMines) [inline EasyCommands/CommandParsers/ParameterParsers.cs:425]: nit: I think we can do the following:

```
            if (token.isExplicitString)
                commandParameters.Add(new VariableCommandParameter(GetStaticVariable(token.original)));
            else if (!token.isString && propertyWords.ContainsKey(token.token))
                commandParameters.AddList(propertyWords[token.token]);
            else 
                commandParameters.Add(new AmbiguousStringCommandParameter(token.original, token.isString, ParseCommandParameters(Tokenize(token.token)).ToArray()));
```

Comment (MerlinofMines) [reply EasyCommands/CommandParsers/ParameterParsers.cs:425]: not sure if the extra tokenization matters.  *shrug*

File: EasyCommands/CommandParsers/ParsingEngine.cs [hunk 11]
```diff
@@ -40,30 +40,16 @@ partial class Program {
...
-                }
+                    var nameAndParams = Tokenize(commandStrings[i].Remove(0, 1).Trim());
+                    var functionName = nameAndParams[0].original;
 
-                foreach (int i in functionIndices) {
-                    int startingLineNumber = i + 1 + implicitMainOffset;
-                    String functionString = commandStrings[i].Remove(0, 1).Trim();
-                    List<Token> nameAndParams = Tokenize(functionString);
-                    String functionName = nameAndParams[0].original;
+                    functions[functionName] = new FunctionDefinition(functionName, nameAndParams.Skip(1).Select(t => t.original).ToList());
```
Comment (MerlinofMines) [inline EasyCommands/CommandParsers/ParsingEngine.cs:49]: lol glad to have removed this.  i think it was a relic from before dynamic function parsing, where i needed to know the function names up front and so had to do 2 passes.  Glad to simplify this.

**Review by MerlinofMines — APPROVED** [review]
